zkvm/src/types.rs

```diff
@@ -136,6 +139,43 @@ impl ScalarWitness {
     }
 }
 
+impl Neg for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn neg(self) -> ScalarWitness {
+        match self {
+            ScalarWitness::Integer(a) => ScalarWitness::Integer(-a),
+            ScalarWitness::Scalar(a) => ScalarWitness::Scalar(-a),
+        }
+    }
+}
+impl Add for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn add(self, rhs: ScalarWitness) -> ScalarWitness {
+        match (self, rhs) {
+            (ScalarWitness::Integer(a), ScalarWitness::Integer(b)) => {
+                let res = a + b;
+
+                if res > spacesuit::SignedInteger::from(std::u64::MAX) {
+                    return ScalarWitness::Scalar(res.into());
```

Comment:
instead of early return, i'd use if-else:

```rust
if res > max || res < min {
    ScalarWitness::Scalar(res.into())
} else {
    ScalarWitness::Integer(res)
}
```